common/jsonrpc_errors.h

```diff
@@ -3,54 +3,58 @@
  */
 #ifndef LIGHTNING_COMMON_JSONRPC_ERRORS_H
 #define LIGHTNING_COMMON_JSONRPC_ERRORS_H
+
 #include "config.h"
 
+#include <common/errcode.h>
+
 /* Standard errors defined by JSON-RPC 2.0 standard */
-#define JSONRPC2_INVALID_REQUEST	-32600
-#define JSONRPC2_METHOD_NOT_FOUND	-32601
-#define JSONRPC2_INVALID_PARAMS		-32602
+extern const errcode_t JSONRPC2_INVALID_REQUEST;
```

Comment:
This.... just adds some space to our .const sections without reducing the actual .text sections in practice (because the code in .text will need an address to read this variable from).  It also increases the effort to make a new error code, now we have to modify the `.h` *and* a `.c`, and we already know that developers do not make the effort to make a new error code in a single `.h` file already, everybody just keeps reusing `LIGHTNINGD` (which is sad, you are saying `lightningd` itself is an error....).

Suggest instead `static const errcode_t JSONRPC2_INVALID_REQUEST = -32600` which decent compilers should optimize away (i.e. no actual storage is allocated for these variables) if not used by the compilation unit\.  `gdb` will still be able to grok these variables as well.

Reply:
I deliberately avoided `static` variables in the `.h` file because that will define the symbols in every `.o` file (that uses them).

Currently there are no `static` variables defined in header files, but there are 15 `extern` ones:

```
common/bech32.h:119:extern const char bech32_charset[32];
common/bech32.h:120:extern const int8_t bech32_charset_rev[128];
common/configdir.h:9:extern bool deprecated_apis;
common/derive_basepoints.h:160:extern struct secrets *dev_force_channel_secrets;
common/derive_basepoints.h:161:extern struct sha256 *dev_force_channel_secrets_shaseed;
common/dev_disconnect.h:33:extern bool dev_suppress_commit;
common/per_peer_state.h:61:extern bool dev_fast_gossip;
common/status.h:33:extern volatile bool logging_io;
common/utils.h:11:extern secp256k1_context *secp256k1_ctx;
common/utils.h:13:extern const struct chainparams *chainparams;
common/utils.h:63:extern const tal_t *tmpctx;
gossipd/seeker.h:27:extern bool dev_suppress_gossip;
lightningd/lightningd.h:261:extern bool tal_oom_ok;
lightningd/log.h:59:extern struct log *crashlog;
plugins/libplugin.h:19:extern bool deprecated_apis;
```


Done in fb8ab0aa.

Reply:
Seems to be optimized out with `-O2`:

```
zmnscpxj@zmnscpxj-server-2020:~/dvpt/tmp$ cat main.c

static const int FOO = 42;
static const int BAR = 43;

int main() {
        return FOO;
}

zmnscpxj@zmnscpxj-server-2020:~/dvpt/tmp$ gcc -O2 -g main.c
zmnscpxj@zmnscpxj-server-2020:~/dvpt/tmp$ gdb ./a.out
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./a.out...done.
(gdb) p &FOO
Can't take address of "FOO" which isn't an lvalue.
(gdb) p &BAR
Can't take address of "BAR" which isn't an lvalue.
(gdb) p FOO
$1 = 42
(gdb) p BAR
$2 = 43
(gdb) quit
```

Reply:
Also optimized away in `gcc -O1`.  At `gcc -O0` the variables are instantiated and it is possible to get their address from `gdb`.  This is `gcc --version` `gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609`, which seems somewhat stable; latest GCC is 9.2, it seems unlikely to me that this will be reverted in future GCC.

Reply:
Looking at the assembly from `gcc -S`, well it looks like the `.debug_info` section is populated with duplicates from each compilation unit, so you have a point there, but debug information is not loaded during normal execution and the generated binary code is shorter with `static` than with `extern` (because no need for an externally-resolved address).  Installer can always `strip` debug information from a binary later if size of the debugging info is an issue.

Reply:
I see `static` wins :)